### PR TITLE
Feat/3411  add span wrapping mention

### PR DESCRIPTION
**Pull request by gbonaspetti**
#3411 

## Checkpoints

These points must be checked before merging. Please don't edit them out.

**For developers**

- [x] If relevant, manual tests have been done to ensure the stability of the whole application and that the involved feature works
- [x] The original issue is up to date w.r.t the latest discussions and contains a short summary of the implemented solution
- [x] Automated tests covering the feature or the fix, have been written, deemed irrelevant (give the reason), or an issue has been created to implement the test (give the link)

**For code reviewers**

- [x] The code is clear enough
- [x] If there are FIXMEs in the code, related issues are mentioned in the FIXME
- [x] If there are TODOs, NOTEs or HACKs in code, the date and the developer initials are present

**For testers**

- [x] Manual, quality tests have been done

To test:
- open your browser console (ctrl + shift + I - Console)
- make mentions at comments and html documents
- see if you have a `mention.created` TLM

### CODE REVIEW

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 1]
```diff
@@ -30,7 +32,7 @@ import {
   TLM_ENTITY_TYPE as TLM_ET,
   TLM_SUB_TYPE as TLM_ST,
   TracimComponent,
-  getCurrentContentVersionNumber
+  wrapMentionInSpanTag
```
Comment (raphj) [inline frontend_app_html-document/src/container/HtmlDocument.jsx:35]:
```suggestion
  wrapMentionsInSpanTags
```

There can be several mentions, right? A plural seems therefore more correct.

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 2]
```diff
@@ -409,14 +411,26 @@ export class HtmlDocument extends React.Component {
   handleSaveHtmlDocument = async () => {
     const { state, props } = this
 
+    let newDocumentForApiWithMention
+    try {
+      newDocumentForApiWithMention = wrapMentionInSpanTag(state.content.raw_content)
+    } catch (e) {
+      if (e === EXCEPTION_MENTION_PARSING) {
```
Comment (raphj) [inline frontend_app_html-document/src/container/HtmlDocument.jsx:418]: This means you raise a constant?
Wouldn't an Error object (or a class inheriting from Error) be cleaner? (and then you can test with `instanceof`).
No hard opinion on this, you can skip this comment.

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 3]
```diff
@@ -409,14 +411,26 @@ export class HtmlDocument extends React.Component {
...
+        this.sendGlobalFlashMessage(props.t('Error while detecting the mentions'))
+      } else {
+        this.sendGlobalFlashMessage(props.t('Error while saving new version'))
```
Comment (raphj) [inline frontend_app_html-document/src/container/HtmlDocument.jsx:421]:
```suggestion
        this.sendGlobalFlashMessage(props.t('Error while saving the new version'))
```
(and translation strings)

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 4]
```diff
@@ -409,14 +411,26 @@ export class HtmlDocument extends React.Component {
   handleSaveHtmlDocument = async () => {
     const { state, props } = this
 
+    let newDocumentForApiWithMention
```
Comment (raphj) [inline frontend_app_html-document/src/container/HtmlDocument.jsx:414]: Maybe this name?
```suggestion
    let contentWithWrappedMentions
```

File: frontend_lib/dist/index.html [hunk 5]
```diff
@@ -7,7 +7,7 @@
   <link rel='shortcut icon' href='favicon.ico'>
 
   <link rel="stylesheet" type="text/css" href="./assets/font/font-awesome-4.7.0/css/font-awesome.css">
-  <link href="https://fonts.googleapis.com/css?family=Quicksand:300,400,500,700" rel="stylesheet">
+<!--  <link href="https://fonts.googleapis.com/css?family=Quicksand:300,400,500,700" rel="stylesheet">-->
```
Comment (raphj) [inline frontend_lib/dist/index.html:10]: Leftover? Should this be restored, or even removed?

File: frontend_lib/src/appContentFactory.js [hunk 6]
```diff
@@ -143,8 +145,15 @@ export function appContentFactory (WrappedComponent) {
       // see https://github.com/tracim/tracim/issues/1101
       const newCommentForApi = isCommentWysiwyg ? newComment : Autolinker.link(`<p>${convertBackslashNToBr(newComment)}</p>`)
 
+      let newCommentForApiWithMention
+      try {
+        newCommentForApiWithMention = wrapMentionInSpanTag(newCommentForApi)
+      } catch (e) {
+        return new Promise((resolve, reject) => reject(e))
```
Comment (raphj) [inline frontend_lib/src/appContentFactory.js:152]:
```suggestion
        return new Promise.reject(e)
```

File: frontend_lib/src/appContentFactory.js [hunk 7]
```diff
@@ -143,8 +145,15 @@ export function appContentFactory (WrappedComponent) {
       // see https://github.com/tracim/tracim/issues/1101
       const newCommentForApi = isCommentWysiwyg ? newComment : Autolinker.link(`<p>${convertBackslashNToBr(newComment)}</p>`)
 
+      let newCommentForApiWithMention
```
Comment (raphj) [inline frontend_lib/src/appContentFactory.js:148]: If you went with my suggestion for the variable name in my earlier comment, you might want to change it here too.

File: frontend_lib/src/helper.js [hunk 8]
```diff
@@ -394,3 +395,52 @@ export const getCurrentContentVersionNumber = (appFeatureMode, content, timeline
...
   return timeline.filter(t => t.timelineType === 'revision' && t.hasBeenRead).length
 }
+
+export const wrapMentionInSpanTag = (text) => {
+  try {
+    const textArray = text.split('&nbsp;').join(' ').split(' ')
+
+    const wrappedTextArray = textArray.map(subString => {
+      if (subString.includes('@')) {
+        const indexOfAt = subString.indexOf('@')
```
Comment (raphj) [inline frontend_lib/src/helper.js:405]:
```suggestion
      const indexOfAt = subString.indexOf('@')
        if (indexOfAt > -1) {
```

File: frontend_lib/src/helper.js [hunk 9]
```diff
@@ -394,3 +395,52 @@ export const getCurrentContentVersionNumber = (appFeatureMode, content, timeline
   if (appFeatureMode === APP_FEATURE_MODE.REVISION) return content.number
   return timeline.filter(t => t.timelineType === 'revision' && t.hasBeenRead).length
 }
+
+export const wrapMentionInSpanTag = (text) => {
+  try {
+    const textArray = text.split('&nbsp;').join(' ').split(' ')
```
Comment (raphj) [inline frontend_lib/src/helper.js:401]:
```suggestion
    const textArray = text.split(/&nbsp;| /)
```
What about \t or other white chars ? maybe use `\s` ?
With this code, you are replacing non breaking space with regular space at the end. This does not seem right.

I also think that this parsing code is a bit fragile: what happens if there is something that looks like a mention in an HTML attribute? I think you should really use a DOMParser and then loop over the text nodes.

Also, what about using a regex with a back reference to detect the space without splitting the string and potentially mangling special white spaces?
`&nbsp;` is not the only entity for spaces. 

What about `<p>@mention</p>`?

File: yarn.lock [hunk 10]
```diff
@@ -11185,10 +11185,10 @@ fsevents@^1.2.7:
 
 "resolve@patch:resolve@^1.10.0#builtin<compat/resolve>, resolve@patch:resolve@^1.10.1#builtin<compat/resolve>, resolve@patch:resolve@^1.11.0#builtin<compat/resolve>, resolve@patch:resolve@^1.12.0#builtin<compat/resolve>, resolve@patch:resolve@^1.13.1#builtin<compat/resolve>, resolve@patch:resolve@^1.17.0#builtin<compat/resolve>, resolve@patch:resolve@^1.3.2#builtin<compat/resolve>":
   version: 1.17.0
-  resolution: "resolve@patch:resolve@npm%3A1.17.0#builtin<compat/resolve>::version=1.17.0&hash=e7677c"
+  resolution: "resolve@patch:resolve@npm%3A1.17.0#builtin<compat/resolve>::version=1.17.0&hash=3388aa"
   dependencies:
     path-parse: ^1.0.6
-  checksum: 76954aad72cd6ba127db4fcb3354a88be83f532ba7d58e166f776136bbeded410e0f3db86737fe0f3b4714627835bde078fa662887362336aaffb4bb7002a847
+  checksum: 99df12ab341fb8e96c6bce2ea521d3a1294c5c56f85524369808f7294fbf6f01d6f6c7f4408a1f6cd4fb5762b800a889629ff64abc08f12bcc8c3553e3564ab8
```
Comment (raphj) [inline yarn.lock:11191]: is this chunk intentionally committed?

File: frontend_app_html-document/src/container/HtmlDocument.jsx (hunk 2 above)
Comment (grignards) [reply frontend_app_html-document/src/container/HtmlDocument.jsx:418]: I agree with @raphj here: the `Error` object is more appropriate as it also includes a message + relevant information to the raised value.

File: frontend_app_html-document/i18next.scanner/en/translation.json [hunk 11]
```diff
@@ -10,6 +10,8 @@
   "Edit": "Edit",
   "A content with the same name already exists": "A content with the same name already exists",
   "Error while creating document": "Error while creating document",
+  "Error while detecting the mentions": "Error while detecting the mentions",
+  "Error while saving new comment": "Error while saving new comment",
```
[the remark this answers is not in the document]
Comment (gbonaspetti) [reply frontend_app_html-document/i18next.scanner/en/translation.json]: Changed

File: frontend_app_html-document/i18next.scanner/fr/translation.json [hunk 12]
```diff
@@ -10,6 +10,8 @@
   "Edit": "Modifier",
   "A content with the same name already exists": "Un contenu avec le même nom existe déjà",
   "Error while creating document": "Erreur lors de la création du document",
+  "Error while detecting the mentions": "Erreur lors de la détection des mentions",
+  "Error while saving new comment": "Erreur lors de l'enregistrement d'un nouveau commentaire",
```
[the remark this answers is not in the document]
Comment (gbonaspetti) [reply frontend_app_html-document/i18next.scanner/fr/translation.json]: Changed

File: frontend_app_html-document/src/container/HtmlDocument.jsx (hunk 1 above)
Comment (gbonaspetti) [reply frontend_app_html-document/src/container/HtmlDocument.jsx]: Changed

File: frontend_app_html-document/src/container/HtmlDocument.jsx (hunk 4 above)
Comment (gbonaspetti) [reply frontend_app_html-document/src/container/HtmlDocument.jsx:414]: I prefer the name old name, is more specific

File: frontend_app_html-document/src/container/HtmlDocument.jsx (hunk 3 above)
Comment (gbonaspetti) [reply frontend_app_html-document/src/container/HtmlDocument.jsx]: Changed

File: frontend_app_html-document/src/container/HtmlDocument.jsx (hunk 2 above)
Comment (gbonaspetti) [reply frontend_app_html-document/src/container/HtmlDocument.jsx]: Changed

File: frontend_lib/src/appContentFactory.js (hunk 6 above)
Comment (gbonaspetti) [reply frontend_lib/src/appContentFactory.js]: Changed

File: frontend_lib/src/helper.js (hunk 8 above)
Comment (gbonaspetti) [reply frontend_lib/src/helper.js]: Changed

File: frontend_lib/src/helper.js (hunk 9 above)
Comment (gbonaspetti) [reply frontend_lib/src/helper.js]: Changed

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 13]
```diff
@@ -415,11 +414,7 @@ export class HtmlDocument extends React.Component {
     try {
       newDocumentForApiWithMention = wrapMentionsInSpanTags(state.content.raw_content)
     } catch (e) {
-      if (e === EXCEPTION_MENTION_PARSING) {
-        this.sendGlobalFlashMessage(props.t('Error while detecting the mentions'))
-      } else {
-        this.sendGlobalFlashMessage(props.t('Error while saving the new version'))
-      }
+      this.sendGlobalFlashMessage(e.message ? e.message : props.t('Error while saving the new version'))
```
Comment (raphj) [inline frontend_app_html-document/src/container/HtmlDocument.jsx:417]: Suggestion, but it is a matter of taste.

```suggestion
      this.sendGlobalFlashMessage(e.message || props.t('Error while saving the new version'))
```

Note for myself: `e.message` is already localized.

File: frontend_app_html-document/src/container/HtmlDocument.jsx [hunk 14]
```diff
@@ -473,11 +468,7 @@ export class HtmlDocument extends React.Component {
     try {
       props.appContentSaveNewComment(state.content, state.timelineWysiwyg, state.newComment, this.setState.bind(this), state.config.slug)
     } catch (e) {
-      if (e === EXCEPTION_MENTION_PARSING) {
-        this.sendGlobalFlashMessage(props.t('Error while detecting the mentions'))
-      } else {
-        this.sendGlobalFlashMessage(props.t('Error while saving the comment'))
-      }
+      this.sendGlobalFlashMessage(e.message ? e.message : props.t('Error while saving the comment'))
```
Comment (raphj) [inline frontend_app_html-document/src/container/HtmlDocument.jsx:471]:
```suggestion
      this.sendGlobalFlashMessage(e.message || props.t('Error while saving the comment'))
```

File: frontend_lib/src/helper.js [hunk 15]
```diff
@@ -396,71 +395,62 @@ export const getCurrentContentVersionNumber = (appFeatureMode, content, timeline
...
+                ? fragment.appendChild(document.createTextNode(value.substring(mentionIndex + mention.length)))
+                : fragment.appendChild(document.createTextNode(''))
+
+              lastMentionIndex = mentionIndex + mention.length - 1
+            })
+            childNodesList[i].replaceWith(fragment)
+            i = i + 3 * mentionsInThisNode.length
+          } else i++
+        } else {
+          if (!(node.nodeName === 'SPAN' && node.className === 'mention') && node.childNodes.length > 0) depthFirstSearchAndMentionAnalysis(node.childNodes)
```
Comment (raphj) [inline frontend_lib/src/helper.js:444]: Beware. Should Tracim ever migrate to XHTML, nodeName would become lower case. I'd therefore suggest the following to handle both cases:

```suggestion
          if (!(node.nodeName.toLowerCase() === 'span' && node.className === 'mention') && node.childNodes.length > 0) depthFirstSearchAndMentionAnalysis(node.childNodes)
```
You can also leave out `&& node.childNodes.length > 0`, your code should handle this case (that should not happen anyway) nicely.

File: frontend_lib/src/helper.js [hunk 16]
```diff
@@ -396,71 +395,62 @@ export const getCurrentContentVersionNumber = (appFeatureMode, content, timeline
...
+            mentionsInThisNode.forEach((mention, i) => {
+              mentionIndex = value.indexOf(mention, lastMentionIndex)
+
+              let mentionWithSpan = document.createElement('span')
+              mentionWithSpan.className = 'mention'
+              mentionWithSpan.id = `mention-${uuidv4()}`
+              mentionWithSpan.textContent = mention
+
+              mentionIndex === 0
+                ? fragment.appendChild(document.createTextNode(''))
```
Comment (raphj) [inline frontend_lib/src/helper.js:429]: Is an empty text node needed?
I'd suggest adding nothing to the fragment if it works.

(I understand that it's to ensure that there are 3 elements, but... :-) you could use a counter)

File: frontend_lib/src/helper.js [hunk 17]
```diff
@@ -396,71 +395,62 @@ export const getCurrentContentVersionNumber = (appFeatureMode, content, timeline
...
+              fragment.appendChild(mentionWithSpan)
+
+              mentionsInThisNode.length - 1 === i
+                ? fragment.appendChild(document.createTextNode(value.substring(mentionIndex + mention.length)))
+                : fragment.appendChild(document.createTextNode(''))
+
+              lastMentionIndex = mentionIndex + mention.length - 1
+            })
+            childNodesList[i].replaceWith(fragment)
+            i = i + 3 * mentionsInThisNode.length
```
Comment (raphj) [inline frontend_lib/src/helper.js:441]:
```suggestion
            i += 3 * mentionsInThisNode.length
```
Could you put a comment explaining the 3 if decide to not apply my previous suggestion?

File: frontend_lib/src/helper.js [hunk 18]
```diff
@@ -396,71 +395,62 @@ export const getCurrentContentVersionNumber = (appFeatureMode, content, timeline
...
+
+              mentionIndex === 0
+                ? fragment.appendChild(document.createTextNode(''))
+                : fragment.appendChild(document.createTextNode(value.substring(lastMentionIndex, mentionIndex)))
+
+              fragment.appendChild(mentionWithSpan)
+
+              mentionsInThisNode.length - 1 === i
+                ? fragment.appendChild(document.createTextNode(value.substring(mentionIndex + mention.length)))
+                : fragment.appendChild(document.createTextNode(''))
```
Comment (raphj) [inline frontend_lib/src/helper.js:436]: same here

File: frontend_lib/test/helper.spec.js [hunk 19]
```diff
@@ -375,8 +374,12 @@ describe('helper.js', () => {
...
 
     describe('if the source is NULL', () => {
-      it('should throw an EXCEPTION_MENTION_PARSING exception', () => {
-        expect(() => wrapMentionsInSpanTags(null)).to.throw(EXCEPTION_MENTION_PARSING)
+      it('should throw an error exception', () => {
+        try {
+          wrapMentionsInSpanTags(null)
+        } catch (e) {
+          expect(typeof e).to.equal('Error')
+        }
```
Comment (raphj) [inline frontend_lib/test/helper.spec.js:382]: I'm not sure checking that `e` is actually an error.
But there is something fishy here: `typeof new Error()` is `'object'`. If you really want to test that e is an instance of error, (e instanceof Error).should.equal(true)

**Review by raphj — CHANGES_REQUESTED** [review]

File: frontend/src/util/tinymceInit.js [hunk 20]
```diff
@@ -76,10 +77,19 @@ import i18n from './i18n.js'
       height: '100%',
       setup: function ($editor) {
         $editor.on('init', function (e) {
+          const id = uniqueId()
+          if ($editor.getBody()) {
+            $editor.dom.add($editor.getBody(), 'p', { id: id }, '')
+            $editor.selection.select($editor.dom.select('p' + id)[0])
+          }
```
[the remark this answers is not in the document]
Comment (gbonaspetti) [reply frontend/src/util/tinymceInit.js:85]: Changed

File: frontend/src/util/tinymceInit.js [hunk 21]
```diff
@@ -76,10 +77,20 @@ import i18n from './i18n.js'
       height: '100%',
       setup: function ($editor) {
         $editor.on('init', function (e) {
+          // INFO - GB - 2020-08-24 - The manipulation below is a hack to add a <p> tag in the end of the text in order to start the text outside the other existents tags
```
Comment (raphj) [inline frontend/src/util/tinymceInit.js:80]:
```suggestion
          // INFO - GB - 2020-08-24 - The manipulation below is a hack to add a <p> tag at the end of the text in order to start the text outside the other existing tags
```

File: frontend_app_file/src/container/File.jsx [hunk 22]
```diff
@@ -436,7 +436,11 @@ export class File extends React.Component {
 
   handleClickValidateNewCommentBtn = () => {
     const { props, state } = this
-    props.appContentSaveNewComment(state.content, state.timelineWysiwyg, state.newComment, this.setState.bind(this), state.config.slug)
+    try {
+      props.appContentSaveNewComment(state.content, state.timelineWysiwyg, state.newComment, this.setState.bind(this), state.config.slug)
+    } catch (e) {
+        this.sendGlobalFlashMessage(e.message ? e.message : props.t('Error while saving the comment'))
```
Comment (raphj) [inline frontend_app_file/src/container/File.jsx:442]:
```suggestion
        this.sendGlobalFlashMessage(e.message || props.t('Error while saving the comment'))
```

File: frontend_app_html-document/src/container/HtmlDocument.jsx (hunk 14 above)
Comment (raphj) [reply frontend_app_html-document/src/container/HtmlDocument.jsx:474]: I see that you applied this suggestion elsewhere and resolved this one without applying it. Pick the one you prefer but I suggest picking the same everywhere :-)

File: frontend_app_thread/src/container/Thread.jsx [hunk 23]
```diff
@@ -285,7 +285,11 @@ export class Thread extends React.Component {
 
   handleClickValidateNewCommentBtn = async () => {
     const { props, state } = this
-    props.appContentSaveNewComment(state.content, state.timelineWysiwyg, state.newComment, this.setState.bind(this), state.config.slug)
+    try {
+      props.appContentSaveNewComment(state.content, state.timelineWysiwyg, state.newComment, this.setState.bind(this), state.config.slug)
+    } catch (e) {
+      this.sendGlobalFlashMessage(e.message ? e.message : props.t('Error while saving the comment'))
```
Comment (raphj) [inline frontend_app_thread/src/container/Thread.jsx:291]:
```suggestion
      this.sendGlobalFlashMessage(e.message || props.t('Error while saving the comment'))
```

File: frontend_lib/src/helper.js [hunk 24]
```diff
@@ -393,6 +393,70 @@ export const getCurrentContentVersionNumber = (appFeatureMode, content, timeline
...
+    const parsedText = parser.parseFromString(text, 'text/html')
+
+    const depthFirstSearchAndMentionAnalysis = childNodesList => {
+      let childNodesListCopy = [...childNodesList]
+      let i = 0
+
+      childNodesListCopy.forEach((node) => {
+        let value = node.nodeValue
+
+        if(node.nodeName === '#text' && value.includes('@')) {
```
Comment (raphj) [inline frontend_lib/src/helper.js:410]:
```suggestion
        if (node.nodeName === '#text' && value.includes('@')) {
```

**Review by raphj — CHANGES_REQUESTED** [review]
Minor stuff and we'll be good to go :-)

File: frontend_app_html-document/src/container/HtmlDocument.jsx (hunk 14 above)
Comment (gbonaspetti) [reply frontend_app_html-document/src/container/HtmlDocument.jsx:474]: Sorry, I agree with you, the resolve it's a mistake.

**Review by raphj — APPROVED** [review]

**Comment by inkhey** [comment]
tested on branch.

**Pull request merged**